### ISSUE TITLE
Image upload support

### ISSUE THREAD

**Issue by maxyou** [andrerpena/react-mde#189]
Now I'm reading imageCommand.tsx. 
Before I try to do it, I hope to know if it's possible to add callback/api for upload image in current project framework?

**Comment by andrerpena** [comment]
It is not possible but I have the intention to support that. I will update this thread as soon as I start working on it

**Comment by andrerpena** [comment]
I am working on mentions now. This will be the next one

**Comment by aaronhayes** [comment]
@andrerpena  Just wondering how you are planning on implementing this? Using some sort of uploader such as [Uppy](https://uppy.io/)?

**Comment by andrerpena** [comment]
Hey @aaronhayes. No.. I intend to call a callback and expect a promise that should resource to the text that I should insert in the cursor position. I also intend to update a status message saying that the update is in progress, and disable the textarea while the promise resolves. This way users can do whatever they want.

However, I can investigate this library, make sure it would work with it and suggest it in the docs if I have a good experience. 

I'm working on some styling improvements and I will go into this feature next.

**Comment by aaronhayes** [comment]
@andrerpena that makes sense. I think your approach is the best option for supporting most use case. 

It would be cool to have support for 3rd party uploads to deal with importing from cloud storage, handling resumable uploads, etc. automatically. Perhaps investigate after callback/promise approach 👍

**Comment by deligent-ant** [comment]
Image upload support？ is ready???  how to use

**Comment by deligent-ant** [comment]
i solved it by change the defult icon of 'image' to my custom button, by this button,i can do anything  i do when the icon click.by the way,i remove the default command of image.part of my code
```js
import { getDefaultCommands } from 'react-mde/lib/js/commands'
import { SvgIcon } from 'react-mde/lib/js/icons'

function customeCommandsPlugin() {
  const commands = getDefaultCommands()
  const rowImgCommand = getRowImgCommand()

  //remove the default command of image
  commands[1].commands = commands[1].commands.filter(
    item => item.name != 'image'
  )

  commands[1].commands.push(customeImgCommand(), rowImgCommand)
  const getIcon = name => {
    if (name.indexOf('c_') !== 0) {
      return <SvgIcon icon={name} />
    } else {
      return COMMANDS_NAME[name] // my cunstom button
    }
  }
  return [commands, getIcon]
}
```

**Comment by phanatuan** [comment]
Is there any update on the upload functionality? Thanks :)

**Comment by n1ru4l** [comment]
@phanatuan Check the code example here: https://github.com/andrerpena/react-mde/pull/227#issue-402332837

**Comment by phanatuan** [comment]
Hi @n1ru4l , 

Thanks for the insight. Your code example works well! I have a bit of problem to customize your example to my workflow: 

A bit about my intended flow for image upload: 

1) User clicks on the Image Upload Icon
2) A modal appears. 
3) User selects either A) existing image (in the system) or B) call API to upload image to S3 - Both options returns an `imageURL` string. 
4) Editor will show `[Uploading <imageName>]` when image is being uploaded. After upload, it will show the markdown `![](imgURL)` 

Here are the points I'm stuck (referring to your [code example](https://github.com/andrerpena/react-mde/pull/227)): 

1) Where is the `selectString` defined? I tried add `api.selectString` but it also does not work. Please see the screenshot

![image](https://user-images.githubusercontent.com/6071142/80297510-bf7db500-87ad-11ea-8189-b860349762c6.png)

2) Your `fileInputRef` works well for the hidden `input` file. How could I make it work for my use case (open Modal => choose image => returns URL)? 

Thank you so much! 

Tuan.

**Comment by n1ru4l** [comment]
@phanatuan https://github.com/dungeon-revealer/dungeon-revealer/pull/370/commits/65d1d7f0f734da12f79c550501e484a57c6ed73a#diff-fe6cda5700d3c911d443751cf211b56eR9-R22


> Your fileInputRef works well for the hidden input file. How could I make it work for my use case (open Modal => choose image => returns URL)?

Open a Modal instead of triggering the click on the input file?

I am also currently writing a blog post regarding this, so if you wait a bit I can share it with you.

**Comment by andrerpena** [comment]
Good news is that I got Cmd+V image upload support working on my branch. 
https://github.com/andrerpena/react-mde/pull/243
I don't plan to have a UI for image upload because I am planning for feature parity with Github.
Tomorrow I will review the breaking changes for 10.0.0 (Shouldn't be many), create the Github release, update the documentation and publish the new 10.0.0 version

**Comment by serranoarevalo** [comment]
@andrerpena Awesome news! Thanks a lot!
Will it support drag and drop as well?

**Comment by andrerpena** [comment]
@serranoarevalo :( Not on this release, but it makes sense to add it. Now that the `save-image` command is already implemented, it should be much easier to just add drag and drop support.

**Comment by andrerpena** [comment]
By the way.. 10.0.3 is already there.. Documentation is still a bit lacky because I didnt have time to update everything.. but the Codesandbox demos on the top of the Readme should get you going

**Comment by serranoarevalo** [comment]
@andrerpena  Thanks! I will check it out. Thank you for your work!

**Comment by jacargentina** [comment]
@andrerpena will you look at a PR for drag and drop support?

**Comment by jacargentina** [comment]
@andrerpena Available as PR on #268
